specs/governance/alligator.md

```diff
@@ -0,0 +1,389 @@
+MUST only migrate the `_account` if it has not been migrated yet.
+
+```solidity
+function migrateAccount(address _account) external
+```
+
+### Getters
+
+For backwards compatibility, the `Alligator` MUST implement all public getter functions of the
+[`GovernanceToken`](gov-token.md) related to delegation and voting power. These functions MUST be used by the
+[`GovernanceToken`](gov-token.md) when an account has been been migrated to the `Alligator` contract. Otherwise,
+the [`GovernanceToken`](gov-token.md) MUST use its own state.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
+
+```solidity
+function getPastTotalSupply(uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `subdelegations`
+
+Retrieves the subdelegations for a given user and delegatee.
+
+```solidity
+function subdelegations(address _account, address _delegatee) external view returns (SubdelegationRules memory)
+```
+
+### Events
+
+#### `Subdelegation`
+
+MUST trigger when an account subdelegates voting power to another address (delegatee).
+
+```solidity
+event Subdelegation(address indexed account, address indexed delegatee, SubdelegationRules rule)
+```
+
+#### `Subdelegations`
+
+MUST trigger when an account subdelegates voting power to multiple addresses (delegatees).
+
+```solidity
+event Subdelegations(address indexed account, address[] delegatee, SubdelegationRules[] rules)
+```
+
+#### `DelegateVotesChanged`
+
+MUST trigger every time the voting power of a user changes, including when a token transfer occurs or a subdelegation
+is updated.
+
+```solidity
+event DelegateVotesChanged(address indexed delegate, uint256 previousBalance, uint256 newBalance)
+```
+
+## Storage
+
+The `Alligator` contract MUST be able to store subdelegation rules and checkpoints. These storage variables MUST be
+defined as in the [`GovernanceToken`](gov-token.md) and use the same types:
+
+```solidity
+// Addresses that had their delegation state migrated from the `GovernanceToken` to the `Alligator`.
+mapping(address => bool) public migrated;
+
+// Subdelegation rules for an account and delegatee.
+mapping(address => mapping(address => SubdelegationRules)) internal _subdelegations;
```

Comment:
I think we need a redundant backwards `subdelegations` mapping here as well in order to calculate voting power
`mapping(address => mapping(address => SubdelegationRules)) internal _subdelegationsTo;`
`mapping(address => mapping(address => SubdelegationRules)) internal _subdelegationsFrom;`

`voting_power = sum(subdelegationsTo) - sum(subdelegatesionsFrom)`